modules/dag-l1/src/main/scala/org/tessellation/dag/l1/domain/transaction/ContextualTransactionValidator.scala

```diff
@@ -0,0 +1,282 @@
+      ): F[ContextualTransactionValidationErrorOr[Signed[Transaction]]] =
+        txs match {
+          case Some(txs) =>
+            signedTx.toHashed.map(resolveConflict(_, txs)).map {
+              case NoConflict | CanOverride          => signedTx.validNec[ContextualTransactionValidationError]
+              case CannotOverride(existing, current) => Conflict(signedTx.ordinal, existing.hash, current.hash).invalidNec
+            }
+          case None => signedTx.validNec[ContextualTransactionValidationError].pure[F]
+        }
+
+      private def validateLastTransactionRef(
+        signedTx: Signed[Transaction],
+        txs: Option[SortedMap[TransactionOrdinal, StoredTransaction]]
+      ): ContextualTransactionValidationErrorOr[Signed[Transaction]] = {
+        val lastProcessedTransactionRef = context.getLastProcessedTransactionRef(txs)
+
+        if (signedTx.parent.ordinal < lastProcessedTransactionRef.ordinal)
+          ParentOrdinalLowerThenLastProcessedTxOrdinal(signedTx.parent.ordinal, lastProcessedTransactionRef.ordinal).invalidNec
+        else if (signedTx.parent === lastProcessedTransactionRef) {
+          signedTx.validNec
+        } else {
+          val hasValidParent = txs.exists(_.exists {
+            case (parentOrdinal, parentTx) => parentTx.ref === signedTx.parent && parentOrdinal > lastProcessedTransactionRef.ordinal
+          })
+
+          if (hasValidParent) signedTx.validNec else HasNoMatchingParent(signedTx.parent.hash).invalidNec
+        }
+      }
+
+      private def validateNotLocked(
+        signedTx: Signed[Transaction]
+      ): ContextualTransactionValidationErrorOr[Signed[Transaction]] =
+        Validated
+          .condNec[ContextualTransactionValidationError, Signed[Transaction]](
+            !lockedAddresses.contains(signedTx.source),
+            signedTx,
+            LockedAddressError(signedTx.source)
+          )
+
+      private def recursivelyValidateCap(
+        minFee: TransactionFee,
+        ordinalDistanceForBaseBalance: Long,
+        baseBalance: Balance,
+        memPool: List[Hashed[Transaction]],
+        incomingTx: Hashed[Transaction],
+        balance: Balance,
+        lastMajorityTransactionOrdinal: SnapshotOrdinal,
+        currentOrdinal: SnapshotOrdinal
+      )(implicit next: Next[SnapshotOrdinal]): Option[Hashed[Transaction]] = {
+        @tailrec
+        def loop(
+          memPool: List[Hashed[Transaction]],
+          balance: Balance,
+          lastTxRefOrdinal: SnapshotOrdinal,
+          currentOrdinal: SnapshotOrdinal,
+          totalConsumedCap: Double,
+          toProcess: List[Hashed[Transaction]]
+        ): List[Hashed[Transaction]] = {
+          val distancePassed = currentOrdinal.value - lastTxRefOrdinal.value
+          val transactionsAllowedWithinInterval = balance.value.toDouble / baseBalance.value.toDouble
+          val distanceBetweenTxsNeeded = ordinalDistanceForBaseBalance.toDouble / transactionsAllowedWithinInterval
```

Comment:
For `transactionsAllowedWithinInterval` you are right. It would be nice to just fallback to `0` so no transactions are allowed. 
For the `distanceBetweenTxsNeeded` I think it is fine to leave it because `ordinalDistanceForBaseBalance.toDouble / 0.0` gives `Infinity` which is kinda valid in the context of that functionality:

If I have `Balance(0)` then I need to wait `Infinity` ordinals to make a transaction (which kinda makes sense because balance 0 will never make a transaction). Then `consumedCap` (cap is the remaining space to "fit" the transaction) is then also `Infinity` so I used the whole available cap immediately.

On the other hand keep in mind that `baseBalance` is something we set in config and setting it to 0 doesn't make sense in the same way as setting `timeTriggerInterval` for snapshots to `0` doesn't make sense because it will trigger consensus every `0` seconds 🤷🏻‍♂️ . Maybe it is worth to introduce the type `NonNegBalance` though or get rid of incorrect states at the config level in general.

Anyway, good catch. I will use one of solutions described above.